src/attacks.rs

```diff
@@ -224,35 +224,27 @@ fn count_paths(g: &Graph, s: &HashSet<usize>, length: usize, k: usize) -> (Vec<u
 
     // counting how many incident paths of length d there is for each node
     let mut incidents = Vec::with_capacity(g.size());
-    let mut topk = vec![Pair(0, 0); k];
     // counting the top k node wo have the greatest number of incident paths
     // NOTE: difference with the C# that recomputes that vector separately.
     // Since topk is directly correlated to incidents[], we can compute both
     // at the same time and remove one O(n) iteration.
     g.for_each_node(|&node| {
-        let node_idx = incidents.len();
-        let count = (0..=length)
-            .map(|d| starting_paths[node][d] * ending_paths[node][length - d])
-            .sum();
-        incidents.push(count);
-
-        // find the smaller element in top k
-        let (idx, pair) = topk
-            .iter()
-            .cloned()
-            .enumerate()
-            .min_by_key(|(_, pair)| pair.1)
-            .unwrap();
-
-        // replace if the minimum number of incident paths in topk is smaller
-        // than the one computed for node i in this iteration
-        if pair.1 < count {
-            topk[idx] = Pair(node_idx, count);
-            //println!("topk {:?}", topk);
-        }
+        incidents.push(Pair(node, (0..=length).map(|d| {
+            starting_paths[node][d] * ending_paths[node][length - d]
+        }).sum()));
     });
-    topk.sort_by_key(|p| Reverse(p.1));
-    (incidents, topk)
+
+    let incidents_return = incidents.iter().map(|pair| {pair.1}).collect();
+    incidents.sort_by_key(|pair| pair.1);
+    incidents.reverse();
```

Comment:
I am not 100% sold if this is an optimization or not:
1. It uses 2n memory instead of n (`incidents` and `incidents_return`)
2. It sorts by key the incidents, inducing O(n log(n)) complexity
3. It reverse the incidents, inducing a O(n) complexity

The current code is a bit uglier for sure, but it avoids the sorting. What do you think ?

Note that in the optimization I'm doing now (issue #15 ), I'm re-inserting the sorting phase, but that'll land in a different function.

Reply:
I leave it up to you how to best adapt this to the current code, change anything you see fit.

> * It uses 2n memory instead of n (`incidents` and `incidents_return`)

I was following the current API, there is definitely redundancy there.



> 3\. It reverse the incidents, inducing a O(n) complexity

Yes, that can be done in a single step.

> The current code is a bit uglier for sure, but it avoids the sorting. What do you think ?

Benchmark it, I got faster results with `greedy_attacks`.